newapi/ooniapi/rate_limit_quotas.py

```diff
@@ -59,78 +112,64 @@ def __init__(
             self._whitelisted_ipaddrs.add(ipaddress.ip_address(ipa))
 
         self.increment_quota_counters(1)
-        self.refresh_quota_counters_if_needed()
+        self.increment_quota_counters_if_needed()
 
     def increment_quota_counters(self, tdelta: int):
-        """Delta: time from previous run in seconds"""
+        """Increment quota counters for every tracked ipaddr. When they exceed
+        the default value simply delete the key"""
         if tdelta <= 0:
             return
 
-        # TODO: use zip()
-        iterable = (
-            (30 * 24, self._ipaddr_limits[0], self._ipaddr_buckets[0]),
-            (7 * 24, self._ipaddr_limits[1], self._ipaddr_buckets[1]),
-            (1 * 24, self._ipaddr_limits[2], self._ipaddr_buckets[2]),
-        )
-        for hours, limit, bucket in iterable:
-            vdelta = limit / hours / 3600 * tdelta
-            to_delete = []
-            for k, v in bucket.items():
-                v += vdelta
-                if v >= limit:
-                    to_delete.append(k)
-                else:
-                    bucket[k] = v
-
-            for k in to_delete:
-                del bucket[k]
-
-    def _generate_bucket_stats(self):
-        periods = ("month", "week", "day")
-        for b, period in zip(self._ipaddr_buckets, periods):
-            size = len(b)
-            metrics.gauge(f"rate-limit-ipaddrs-{period}", size)
-            print(size)
-
-    def refresh_quota_counters_if_needed(self):
+        # the value is stored as interger in milliseconds
+        iterable = zip(self._hours, self._ipaddr_limits, self.ipaddr_buckets)
+        for hours, limit, db in iterable:
+            vdelta = 1000 * tdelta * limit / hours / 3600
+            with self._lmdb._env.begin(db=db, write=True) as txn:
+                i = txn.cursor().iternext()
+                for raw_ipa, raw_val in i:
+                    v = int(raw_val) + vdelta
+                    if v >= limit:
+                        txn.pop(raw_ipa)
+                    else:
+                        txn.put(raw_ipa, str(int(v)).encode())
```

Comment:
Not sure it's worth the trouble, but maybe since we are going to be doing this a lot of times and for a lot of clients, we might be better of storing the big endian byte representation of the int, instead of having to encode/decode the string representation of them.

Reply:
In terms of computational cost it's negligible compared to the rest of the work the API does. I used a string to be able to access the DB directly for debugging but we could indeed store a raw int.